### PR TITLE
DP-108 - activity leaks

### PR DESCRIPTION
**Pull request by yosriz**
#### Main purpose:
Fix leaks (Activities and everything else) sourced at `recovery` module.
#### Technical description:
* `BroadcastManagerImpl` - each register call adds a new
receiver to the Android's LocalBroadcastManager which is static module, while
in unregister we only clear a single receiver, thus leaking the BroadcastManagerImpl
and eventually the activity that its holds in a reference.
* `CreatePassword/RestoreEnterPasswordFragment` - TextWatcherAdapter used Timer that did not cancelled and thus caused a leak
(TimerTask uses Thread internally and this thread kept reference to the callback,
to the holding fragment and so on), Timer is not needed at all, and handler can
be used.
As a "release" method is needed on destory to remove handler callbacks,
that requires keeping TextWatcherAdapter refernce, I removed TextWatcherAdapter
and make everything local.
#### Dilemmas you faced with? 
I think there's no need for nulling out references, that may obscure internal objects that are leaked. it's preferable to let the activity leak in this case, and we have more chance catch it by leak canary, instead of falsly having no activity leaked but internl objects.
#### Tests:
Manually, used LeakCanary and `dumpsys meminfo` as the real source to check after everything closed that no acitvity or view was leaked.
#### Documentation:
N/A

### CODE REVIEW

File: recovery/src/main/java/com/kin/ecosystem/recovery/backup/view/AbstractTextWatcher.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package com.kin.ecosystem.recovery.backup.view;
+
+import android.text.Editable;
+import android.text.TextWatcher;
+
+
+public class AbstractTextWatcher implements TextWatcher {
```
Comment (OrenZak) [inline recovery/src/main/java/com/kin/ecosystem/recovery/backup/view/AbstractTextWatcher.java]: Delete and use the Adapter with Handler inside.

Comment (yosriz) [reply recovery/src/main/java/com/kin/ecosystem/recovery/backup/view/AbstractTextWatcher.java]: fixed :ok_hand:

File: recovery/src/main/java/com/kin/ecosystem/recovery/backup/view/TextWatcherAdapter.java [hunk 2]
```diff
@@ -32,22 +29,16 @@ public void beforeTextChanged(CharSequence s, int start, int count, int after) {
...
-						@Override
-						public void run() {
-							listener.afterTextChanged(editable);
-						}
-					});
-				}
-			},
-			DELAY
-		);
+		mainThreadHandler.postDelayed(new Runnable() {
```
Comment (OrenZak) [inline recovery/src/main/java/com/kin/ecosystem/recovery/backup/view/TextWatcherAdapter.java:33]: you need to cancel the Runnable before each post.

File: recovery/src/main/java/com/kin/ecosystem/recovery/restore/view/RestoreEnterPasswordFragment.java [hunk 3]
```diff
@@ -163,4 +165,11 @@ public void invalidQrError() {
 		contentText.setTextColor(ContextCompat.getColor(getContext(), R.color.kinrecovery_red));
 		password.setFrameBackgroundColor(R.color.kinrecovery_red);
 	}
+
+	@Override
+	public void onDestroy() {
+		super.onDestroy();
+
+		textWatcherAdapter.release();
```
Comment (OrenZak) [inline recovery/src/main/java/com/kin/ecosystem/recovery/restore/view/RestoreEnterPasswordFragment.java:173]: should be called before super.

File: recovery/src/main/java/com/kin/ecosystem/recovery/events/BroadcastManagerImpl.java [hunk 4]
```diff
@@ -31,46 +29,38 @@
...
-		LocalBroadcastManager.getInstance(activity).registerReceiver(receiver, new IntentFilter(actionName));
+		receiversList.add(broadcastReceiver);
+		LocalBroadcastManager.getInstance(activity).registerReceiver(broadcastReceiver, new IntentFilter(actionName));
 	}
 
 	@Override
-	public void unregister() {
-		if (receiver != null && activity != null) {
-			LocalBroadcastManager.getInstance(activity).unregisterReceiver(receiver);
+	public void unregisterAll() {
```
Comment (OrenZak) [inline recovery/src/main/java/com/kin/ecosystem/recovery/events/BroadcastManagerImpl.java:48]: should we add option to delete a listener, and not all of them ?

File: recovery/src/main/java/com/kin/ecosystem/recovery/events/BroadcastManagerImpl.java [hunk 5]
```diff
@@ -31,46 +29,38 @@
 
 	public BroadcastManagerImpl(@NonNull Activity activity) {
 		this.activity = activity;
+		receiversList = new ArrayList<>();
```
Comment (OrenZak) [inline recovery/src/main/java/com/kin/ecosystem/recovery/events/BroadcastManagerImpl.java]: this. receiversList ?

File: recovery/src/main/java/com/kin/ecosystem/recovery/backup/view/CreatePasswordFragment.java [hunk 6]
```diff
@@ -205,4 +212,12 @@ public void onClick(DialogInterface dialog, int which) {
 			.create()
 			.show();
 	}
+
+	@Override
+	public void onDestroy() {
+		super.onDestroy();
+
+		confirmPassTextWatcherAdapter.release();
```
Comment (OrenZak) [inline recovery/src/main/java/com/kin/ecosystem/recovery/backup/view/CreatePasswordFragment.java:220]: before super ?

**Comment by OrenZak** [comment]
sync with dev please

File: recovery/src/main/java/com/kin/ecosystem/recovery/events/BroadcastManagerImpl.java (hunk 4 above)
Comment (yosriz) [reply recovery/src/main/java/com/kin/ecosystem/recovery/events/BroadcastManagerImpl.java:48]: as it's internal mechanism let's add it by demand in the future

File: recovery/src/main/java/com/kin/ecosystem/recovery/backup/view/TextWatcherAdapter.java (hunk 2 above)
Comment (yosriz) [reply recovery/src/main/java/com/kin/ecosystem/recovery/backup/view/TextWatcherAdapter.java:33]: :+1:, sure

File: recovery/src/main/java/com/kin/ecosystem/recovery/restore/view/RestoreEnterPasswordFragment.java (hunk 3 above)
Comment (yosriz) [reply recovery/src/main/java/com/kin/ecosystem/recovery/restore/view/RestoreEnterPasswordFragment.java:173]: why? don't think there is some particular guide regarding that.
From looking at base class (support v4 fragment) super.onDestory merely releasing loader if any (which we don't have, so it does nothing), also looking at `AppCompatActivity` code, there's nothing I can say that must happen at the end. (removing internal handlers callbacks, flagging that activity is destoryed etc.0

File: recovery/src/main/java/com/kin/ecosystem/recovery/events/BroadcastManagerImpl.java (hunk 4 above)
Comment (OrenZak) [reply recovery/src/main/java/com/kin/ecosystem/recovery/events/BroadcastManagerImpl.java:48]: 👍

File: recovery/src/main/java/com/kin/ecosystem/recovery/restore/view/RestoreEnterPasswordFragment.java (hunk 3 above)
Comment (OrenZak) [reply recovery/src/main/java/com/kin/ecosystem/recovery/restore/view/RestoreEnterPasswordFragment.java:173]: Well, i've noticed that could happened that you resources won't be GC'ed.
But if you sure 👍

**Comment by OrenZak** [comment]
@yosriz did you ran a test with `don't keep activities` and do a process restart test ?

**Review by OrenZak — APPROVED** [review]

**Pull request merged**